pkg/rtc/participant.go

```diff
@@ -1283,38 +1292,57 @@ func (p *ParticipantImpl) handleDataMessage(kind livekit.DataPacket_Kind, data [
 	}
 }
 
-func (p *ParticipantImpl) logSelectedCandidatePair(isPrimary bool) {
+func (p *ParticipantImpl) getTransport(isPrimary bool) *PCTransport {
 	pcTransport := p.publisher
 	if (isPrimary && p.SubscriberAsPrimary()) || (!isPrimary && !p.SubscriberAsPrimary()) {
 		pcTransport = p.subscriber
 	}
 
-	sctp := pcTransport.pc.SCTP()
-	if sctp == nil {
-		return
-	}
+	return pcTransport
+}
 
-	transport := sctp.Transport()
-	if transport == nil {
-		return
+func (p *ParticipantImpl) handleICEConnected(isPrimary bool) {
+	pcTransport := p.getTransport(isPrimary)
+	pcTransport.SetICEConnectedAt(time.Now())
+
+	if pair, err := pcTransport.GetSelectedPair(); err != nil {
+		pcTransport.Logger().Errorw("error getting selected ICE candidate pair", err)
+	} else {
+		pcTransport.Logger().Infow("selected ICE candidate pair", "pair", pair)
 	}
+}
 
-	iceTransport := transport.ICETransport()
-	if iceTransport == nil {
-		return
+func (p *ParticipantImpl) handleConnectionFailed(isPrimary bool) {
+	pcTransport := p.getTransport(isPrimary)
+	isShort, duration := pcTransport.IsShortConnection(time.Now())
+	if isShort {
+		// irrespective of which one fails, force TCP on both as the other one might
+		// fail at a different time and cause another disruption
+		pair, err := pcTransport.GetSelectedPair()
+		if err != nil {
+			pcTransport.Logger().Errorw("short ICE connection", err, "duration", duration)
+		} else {
+			pcTransport.Logger().Infow("short ICE connection", "pair", pair, "duration", duration)
+		}
+		/*
+			p.lock.Lock()
+			p.iceConfig.PreferSubTcp = true
+			p.iceConfig.PreferPubTcp = true
+			p.lock.Unlock()
+		*/
```

Comment:
Commented out for now till we can get more data. Also, need to figure out how to accept relay TLS candidates.

But, the logic here is that on any failure (ICE or peer connection), if the connection is short lived, fallback to TCP. In Pion, ICE connection failure is set to 25 seconds (actually 10 seconds for disconnect detection + 25 seconds for failed detection for a total of 35 seconds) and peer connection fails (because DTLS fails) five seconds before ICE connection is declared failed. So, on either failure, run this.

Also, change both peer connections to use TCP although only one might have failed.

Reply:
that sounds good. for relay it's potentially tricky since the client would need:
* UDP host candidates to be sent back
* JoinResponse to indicate client should forceRelay

Reply:
Yeah, need to think about it more.